axelrod/strategies/axelrod_second.py

```diff
@@ -199,3 +202,210 @@ def strategy(self, opponent: Player) -> Action:
+        two_turns_after_good_defection_ratio and the probability values, 
+        as well as sets the value of num_turns_after_good_defection.
+        """
+
+        if self.num_turns_after_good_defection == 2:
+            self.num_turns_after_good_defection = 0
+            self.two_turns_after_good_defection_ratio = (
+                ((self.two_turns_after_good_defection_ratio 
+                * self.two_turns_after_good_defection_ratio_count) 
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - ((self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]]))) 
+                / (self.two_turns_after_good_defection_ratio_count + 1)
+                )
+            self.two_turns_after_good_defection_ratio_count += 1
+        elif self.num_turns_after_good_defection == 1:
+            self.num_turns_after_good_defection = 2
+            self.one_turn_after_good_defection_ratio = (
+                ((self.one_turn_after_good_defection_ratio 
+                * self.one_turn_after_good_defection_ratio_count)
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - (self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]])) 
+                / (self.one_turn_after_good_defection_ratio_count + 1)
+                )
+            self.one_turn_after_good_defection_ratio_count += 1
+   
+    def update_stochastic_state(self, opponent):
+            
+        if (self.current_score[0] / ((len(self.history)) + 1)) >= 2.25:
+                self.probability = (
+                    (.95 - (((self.one_turn_after_good_defection_ratio)
+                    + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
+                    + (1 / (((len(self.history))+1) ** 2))
+                    - (self.dict[opponent.history[-1]] / 4)
+                    )
+                self.probability = round(self.probability, 4)
+                self.score = "good"
+        elif (self.current_score[0] / ((len(self.history)) + 1)) >= 1.75:
+                self.probability = (
+                    (.25 + ((opponent.cooperations + 1) / ((len(self.history)) + 1)))
+                    - (self.opponent_consecutive_defections * .25) 
+                    + ((self.current_score[0] 
+                    - self.current_score[1]) / 100) 
+                    + (4 / ((len(self.history)) + 1))
+                    )
+                self.probability = round(self.probability, 4)
+                self.score = "average"
+
+    def strategy(self, opponent: Player) -> Action:
+
+        self.current_score = compute_final_score(zip(self.history, opponent.history))
+        
+        if len(self.history) == 0:
+            return C
+        else: 
+            self.update_cooperative_state(opponent)
```

Comment:
Before this line, let's go for a:

```
if  self.num_turns_after_good_defection in [1, 2]:
    self.update_cooperative_state(opponent)
    return C
```

This is a nice quick and easy to debug way to say that if we're in a cooperative state, update the cooperative state and then return a `C`. The whole rest of the code can then be cleaned up quite a lot to consider the case of when we're not in a cooperative state.

Reply:
I think I may have misunderstood something here, but how did you come to reason that given `self.num_turns_after_good_defection != 0`, the response would be to `return C`? I found the response to be dependent on the action of the opponent on the last turn (due to the fact that the score may drop below 2.25 and hence allow for double defection).

Reply:
We're both a bit wrong here.

These lines: https://github.com/Axelrod-Python/TourExec/blob/v0.3.0/src/strategies/k67r.f#L29:

```
545   K67R = 0
      IF (ABS(FD - 1.5) .EQ. .5) GOTO 599
```

mean that if FD is 1 or 2 (what `ABS(FD - 1.5) .EQ. .5`) is equivalent to then we go to 599 (which is just a return.

Where I am a bit wrong is that this needs to happen after we update the `cooperative states`. So: if after updating the cooperative states the state is in [1, 2] then return `C`:


```python
self.update_cooperative_state(opponent)
if  self.num_turns_after_good_defection in [1, 2]:
    return C
```

This is essentially what lines 15 to 29 do in the Fortran code.